### PR TITLE
SG Production Account Setup

### PR DESCRIPTION
**Pull request by sinehome**
Setting up the production account for Tinkle Friend (Singapore)

### CODE REVIEW

**Comment by sinehome** [comment]
Plan with the helpline region changed to 'eu-west-1' - **There are 2 errors pertaining to S3 buckets**

Terraform planned the following actions, but then encountered a problem:

  # module.aws.aws_ssm_parameter.aws_region will be updated in-place
  ~ resource "aws_ssm_parameter" "aws_region" {
        id             = "/production/aws/AC0751f021d17d50f0ee5af094acdee7c8/region"
      + insecure_value = (known after apply)
        name           = "/production/aws/AC0751f021d17d50f0ee5af094acdee7c8/region"
        tags           = {
            "Environment" = "production"
            "Name"        = "/production/aws/AC0751f021d17d50f0ee5af094acdee7c8/region"
            "Terraform"   = "true"
        }
      ~ value          = (sensitive value)
      ~ version        = 1 -> (known after apply)
        # (7 unchanged attributes hidden)
    }

  # module.taskRouter.twilio_taskrouter_workspaces_workflows_v1.workflow["master"] will be updated in-place
  ~ resource "twilio_taskrouter_workspaces_workflows_v1" "workflow" {
      ~ configuration            = jsonencode( # whitespace changes
            {
                task_routing = {
                    filters = [
                        {
                            expression           = "1==1 OR channelType =='web' OR isContactlessTask == true"
                            filter_friendly_name = "Tinkle Friend"
                            targets              = [
                                {
                                    expression = "(worker.waitingOfflineContact != true AND ((task.channelType == 'voice' AND worker.channel.chat.assigned_tasks == 0) OR (task.channelType != 'voice' AND worker.channel.voice.assigned_tasks == 0)) AND ((task.transferTargetType == 'worker' AND task.targetSid == worker.sid) OR (task.transferTargetType != 'worker' AND worker.sid != task.ignoreAgent))) OR (worker.waitingOfflineContact == true AND task.targetSid == worker.sid AND task.isContactlessTask == true)"
                                    queue      = "WQb542639cb1c664d43d51902bb28c551a"
                                },
                            ]
                        },
                        {
                            expression           = "channelType=='web' AND e2eTestMode=='true'"
                            filter_friendly_name = "E2E Test"
                            targets              = [
                                {
                                    expression = "(worker.waitingOfflineContact != true AND ((task.channelType == 'voice' AND worker.channel.chat.assigned_tasks == 0) OR (task.channelType != 'voice' AND worker.channel.voice.assigned_tasks == 0)) AND ((task.transferTargetType == 'worker' AND task.targetSid == worker.sid) OR (task.transferTargetType != 'worker' AND worker.sid != task.ignoreAgent))) OR (worker.waitingOfflineContact == true AND task.targetSid == worker.sid AND task.isContactlessTask == true)"
                                    queue      = "WQ055bb97cfb7ffc95d9f88cc270d54fb9"
                                },
                            ]
                        },
                    ]
                }
            }
        )
        id                       = "WS03e8228cfbf5f68aa659cf1676890587/WWd56699a4851ad3f2ab79725005ddc2bf"
        # (4 unchanged attributes hidden)
    }

  # module.taskRouter.twilio_taskrouter_workspaces_workflows_v1.workflow["survey"] will be updated in-place
  ~ resource "twilio_taskrouter_workspaces_workflows_v1" "workflow" {
      ~ configuration            = jsonencode( # whitespace changes
            {
                task_routing = {
                    filters = [
                        {
                            expression           = "isChatCaptureControl==true"
                            filter_friendly_name = "CaptureChannel"
                            targets              = [
                                {
                                    queue = "WQ49206bff439d8969decae81ba96eeadb"
                                },
                            ]
                        },
                    ]
                }
            }
        )
        id                       = "WS03e8228cfbf5f68aa659cf1676890587/WWf473b8d331faa7773d095cd7b2af00e3"
        # (4 unchanged attributes hidden)
    }

Plan: 0 to add, 3 to change, 0 to destroy.
╷
│ Error: reading Amazon S3 (Simple Storage) Bucket (tl-aselo-docs-sg-production): BucketRegionError: incorrect region, the bucket is not in 'eu-west-1' region at endpoint ''
│ 	status code: 301, request id: , host id:
│
│   with module.aws.aws_s3_bucket.docs,
│   on ../../aws/default/main.tf line 28, in resource "aws_s3_bucket" "docs":
│   28: resource "aws_s3_bucket" "docs" {
│
╵
╷
│ Error: reading Amazon S3 (Simple Storage) Bucket (tl-public-chat-sg-prod): BucketRegionError: incorrect region, the bucket is not in 'eu-west-1' region at endpoint ''
│ 	status code: 301, request id: , host id:
│
│   with module.aws.aws_s3_bucket.chat,
│   on ../../aws/default/main.tf line 87, in resource "aws_s3_bucket" "chat":
│   87: resource "aws_s3_bucket" "chat" {
│

**Comment by sinehome** [comment]
New TG Plan following deletion of us-east-1 S3 buckets

Terraform will perform the following actions:

  # module.aws.aws_s3_bucket.chat will be created
  + resource "aws_s3_bucket" "chat" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "tl-public-chat-sg-prod"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)
    }

  # module.aws.aws_s3_bucket.docs will be created
  + resource "aws_s3_bucket" "docs" {
      + acceleration_status         = (known after apply)
      + acl                         = (known after apply)
      + arn                         = (known after apply)
      + bucket                      = "tl-aselo-docs-sg-production"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + object_lock_enabled         = (known after apply)
      + policy                      = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + tags_all                    = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)
    }

  # module.aws.aws_s3_bucket_cors_configuration.chat will be created
  + resource "aws_s3_bucket_cors_configuration" "chat" {
      + bucket = "tl-public-chat-sg-prod"
      + id     = (known after apply)

      + cors_rule {
          + allowed_headers = [
              + "*",
            ]
          + allowed_methods = [
              + "GET",
              + "POST",
              + "PUT",
            ]
          + allowed_origins = [
              + "https://flex.twilio.com",
            ]
          + expose_headers  = []
        }
    }

  # module.aws.aws_s3_bucket_cors_configuration.docs will be created
  + resource "aws_s3_bucket_cors_configuration" "docs" {
      + bucket = "tl-aselo-docs-sg-production"
      + id     = (known after apply)

      + cors_rule {
          + allowed_headers = [
              + "*",
            ]
          + allowed_methods = [
              + "GET",
              + "POST",
              + "PUT",
            ]
          + allowed_origins = [
              + "https://flex.twilio.com",
            ]
          + expose_headers  = []
        }
    }

  # module.aws.aws_s3_bucket_lifecycle_configuration.hrm_export_expiry will be created
  + resource "aws_s3_bucket_lifecycle_configuration" "hrm_export_expiry" {
      + bucket = "tl-aselo-docs-sg-production"
      + id     = (known after apply)

      + rule {
          + id     = "HRM Exported Data Expiration Policy"
          + status = "Enabled"

          + expiration {
              + days                         = 30
              + expired_object_delete_marker = (known after apply)
            }

          + filter {
              + prefix = "hrm-data/"
            }
        }
    }

  # module.aws.aws_s3_bucket_ownership_controls.chat will be created
  + resource "aws_s3_bucket_ownership_controls" "chat" {
      + bucket = "tl-public-chat-sg-prod"
      + id     = (known after apply)

      + rule {
          + object_ownership = "ObjectWriter"
        }
    }

  # module.aws.aws_s3_bucket_ownership_controls.docs will be created
  + resource "aws_s3_bucket_ownership_controls" "docs" {
      + bucket = "tl-aselo-docs-sg-production"
      + id     = (known after apply)

      + rule {
          + object_ownership = "ObjectWriter"
        }
    }

  # module.aws.aws_s3_bucket_public_access_block.chat will be created
  + resource "aws_s3_bucket_public_access_block" "chat" {
      + block_public_acls       = false
      + block_public_policy     = false
      + bucket                  = (known after apply)
      + id                      = (known after apply)
      + ignore_public_acls      = false
      + restrict_public_buckets = false
    }

  # module.aws.aws_s3_bucket_public_access_block.docs will be created
  + resource "aws_s3_bucket_public_access_block" "docs" {
      + block_public_acls       = true
      + block_public_policy     = true
      + bucket                  = (known after apply)
      + id                      = (known after apply)
      + ignore_public_acls      = true
      + restrict_public_buckets = true
    }

  # module.aws.aws_s3_bucket_server_side_encryption_configuration.chat will be created
  + resource "aws_s3_bucket_server_side_encryption_configuration" "chat" {
      + bucket = "tl-public-chat-sg-prod"
      + id     = (known after apply)

      + rule {
          + apply_server_side_encryption_by_default {
              + sse_algorithm = "AES256"
            }
        }
    }

  # module.aws.aws_s3_bucket_server_side_encryption_configuration.docs will be created
  + resource "aws_s3_bucket_server_side_encryption_configuration" "docs" {
      + bucket = "tl-aselo-docs-sg-production"
      + id     = (known after apply)

      + rule {
          + apply_server_side_encryption_by_default {
              + sse_algorithm = "AES256"
            }
        }
    }

  # module.aws.aws_ssm_parameter.aws_region will be updated in-place
  ~ resource "aws_ssm_parameter" "aws_region" {
        id             = "/production/aws/AC0751f021d17d50f0ee5af094acdee7c8/region"
      + insecure_value = (known after apply)
        name           = "/production/aws/AC0751f021d17d50f0ee5af094acdee7c8/region"
        tags           = {
            "Environment" = "production"
            "Name"        = "/production/aws/AC0751f021d17d50f0ee5af094acdee7c8/region"
            "Terraform"   = "true"
        }
      ~ value          = (sensitive value)
      ~ version        = 1 -> (known after apply)
        # (7 unchanged attributes hidden)
    }

  # module.taskRouter.twilio_taskrouter_workspaces_workflows_v1.workflow["master"] will be updated in-place
  ~ resource "twilio_taskrouter_workspaces_workflows_v1" "workflow" {
      ~ configuration            = jsonencode( # whitespace changes
            {
                task_routing = {
                    filters = [
                        {
                            expression           = "1==1 OR channelType =='web' OR isContactlessTask == true"
                            filter_friendly_name = "Tinkle Friend"
                            targets              = [
                                {
                                    expression = "(worker.waitingOfflineContact != true AND ((task.channelType == 'voice' AND worker.channel.chat.assigned_tasks == 0) OR (task.channelType != 'voice' AND worker.channel.voice.assigned_tasks == 0)) AND ((task.transferTargetType == 'worker' AND task.targetSid == worker.sid) OR (task.transferTargetType != 'worker' AND worker.sid != task.ignoreAgent))) OR (worker.waitingOfflineContact == true AND task.targetSid == worker.sid AND task.isContactlessTask == true)"
                                    queue      = "WQb542639cb1c664d43d51902bb28c551a"
                                },
                            ]
                        },
                        {
                            expression           = "channelType=='web' AND e2eTestMode=='true'"
                            filter_friendly_name = "E2E Test"
                            targets              = [
                                {
                                    expression = "(worker.waitingOfflineContact != true AND ((task.channelType == 'voice' AND worker.channel.chat.assigned_tasks == 0) OR (task.channelType != 'voice' AND worker.channel.voice.assigned_tasks == 0)) AND ((task.transferTargetType == 'worker' AND task.targetSid == worker.sid) OR (task.transferTargetType != 'worker' AND worker.sid != task.ignoreAgent))) OR (worker.waitingOfflineContact == true AND task.targetSid == worker.sid AND task.isContactlessTask == true)"
                                    queue      = "WQ055bb97cfb7ffc95d9f88cc270d54fb9"
                                },
                            ]
                        },
                    ]
                }
            }
        )
        id                       = "WS03e8228cfbf5f68aa659cf1676890587/WWd56699a4851ad3f2ab79725005ddc2bf"
        # (4 unchanged attributes hidden)
    }

  # module.taskRouter.twilio_taskrouter_workspaces_workflows_v1.workflow["survey"] will be updated in-place
  ~ resource "twilio_taskrouter_workspaces_workflows_v1" "workflow" {
      ~ configuration            = jsonencode( # whitespace changes
            {
                task_routing = {
                    filters = [
                        {
                            expression           = "isChatCaptureControl==true"
                            filter_friendly_name = "CaptureChannel"
                            targets              = [
                                {
                                    queue = "WQ49206bff439d8969decae81ba96eeadb"
                                },
                            ]
                        },
                    ]
                }
            }
        )
        id                       = "WS03e8228cfbf5f68aa659cf1676890587/WWf473b8d331faa7773d095cd7b2af00e3"
        # (4 unchanged attributes hidden)
    }

Plan: 11 to add, 3 to change, 0 to destroy.

**Review by janorivera — APPROVED** [review]
LGTM

**Comment by sinehome** [comment]
Service Configuration plan

Initializing service configuration for sg

------------------
Environment: production
Helpline Code: sg
Account SID: AC0751f021d17d50f0ee5af094acdee7c8

Plan:
Update attributes.hrm_base_url from https://hrm-production.tl.techmatters.org to https://hrm-production-eu.tl.techmatters.org
Update attributes.aws_region from us-east-1 to eu-west-1

File: webchat/configurations/sg-production.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+
+import { Translations, Configuration, MapHelplineLanguage, ContactType } from '../types';
+import { PreEngagementFormDefinition, EMAIL_PATTERN } from '../src/pre-engagement-form';
+
+const accountSid = 'AC0751f021d17d50f0ee5af094acdee7c8';
```
Comment (janorivera) [inline webchat/configurations/sg-production.ts:20]: Usually for production webchats we  just import the staging one

**Review by janorivera — CHANGES_REQUESTED** [review]

**Pull request merged**